src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -413,39 +417,58 @@ internal static IEnumerable<PropertyInfo> GetPropertiesOnType(Type type, bool de
         /// Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.
         /// </summary>
         /// <param name="type">Type in question.</param>
+        /// <param name="context">The context with the client model</param>
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
-        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
+        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, DataServiceContext context)
         {
             bool hasProperties;
-            return GetKeyPropertiesOnType(type, out hasProperties);
+            return GetKeyPropertiesOnType(type, context, out hasProperties);
         }
 
         /// <summary>
         /// Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.
         /// </summary>
         /// <param name="type">Type in question.</param>
+        /// <param name="context">The context with the client model</param>
         /// <param name="hasProperties">true if <paramref name="type"/> has any (declared or inherited) properties; otherwise false.</param>
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
-        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
+        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, DataServiceContext context, out bool hasProperties)
         {
             if (CommonUtil.IsUnsupportedType(type))
             {
                 throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
             }
-
             string typeName = type.ToString();
             IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
+            List<string> DeclaredKeys = new List<string>();
+            if (!customAttributes.Any())

```

Comment:
>f (!customAttributes.Any()) [](start = 13, length = 27)

why test attribute if you have the model to retrieve the keys?

Reply:
The first check checks whether one used the Key Attribute annotation on the key property when creating their model, If not then get the keys from the model.